pkg/stub/handler.go

```diff
@@ -74,7 +74,74 @@ func (h *Handler) EnsureDefaultClusterIngress() error {
 	return fmt.Errorf("creating default cluster ingress %s/%s: %v", ci.Namespace, ci.Name, err)
 }
 
-func (h *Handler) syncIngressUpdate(ci *ingressv1alpha1.ClusterIngress) error {
+// Reconcile performs a full reconciliation loop for ingress, including
+// generalized setup and handling of all clusteringress resources in the
+// operator namespace.
+func (h *Handler) reconcile() error {
+	// Ensure we have all the necessary scaffolding on which to place router
+	// instances.
+	err := h.ensureRouterNamespace()
+	if err != nil {
+		return err
+	}
+
+	// Find all clusteringresses.
+	ingresses := &ingressv1alpha1.ClusterIngressList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterIngress",
+			APIVersion: "ingress.openshift.io/v1alpha1",
+		},
+	}
+	err = sdk.List(h.namespace, ingresses, sdk.WithListOptions(&metav1.ListOptions{}))
+	if err != nil {
+		return fmt.Errorf("failed to list clusteringresses: %v", err)
+	}
+
+	// Reconcile all the ingresses.
+	errors := []error{}
+	for _, ingress := range ingresses.Items {
+		// Handle deleted ingress.
+		// TODO: Assert/ensure that the ingress has a finalizer so we can reliably detect
+		// deletion.
+		if ingress.DeletionTimestamp != nil {
+			// Destroy any router associated with the clusteringress.
+			err := h.ensureRouterDeleted(&ingress)
+			if err != nil {
+				errors = append(errors, fmt.Errorf("couldn't delete clusteringress %q: %v", ingress.Name, err))
+				continue
+			}
+			// Clean up the finalizer to allow the clusteringress to be deleted.
+			ingress.Finalizers = RemoveString(ingress.Finalizers, "ingress.openshift.io/default-cluster-ingress")
+			err = sdk.Update(&ingress)
```

Comment:
Once we update ingress object, don't we see another event for the same object? if yes, that will have DeletionTimestamp but no finalizers section?

Reply:
Yes, and although that still works now, we're doing an additional pointless update. So I think we're at least still correct/reentrant/idempotent... I'll fix

Reply:
Fixed